### ISSUE TITLE
Issue with dropping off/losing logs when pushing to cloudwatch at the end of program execution

### ISSUE THREAD

**Issue by dev-usa** [boxfuse/cloudwatchlogs-java-appender#12]
Hi boxfuse team - 

I tried creating a generic version of cloudwatch logs appender using boxfuse's appender as a starting point. I faced several issues during the process - one of them being that - some of the logs are dropped off and are not fed to CloudWatch. This is because the while loop in the run() method of the CloudwatchLogsLogEventPutter.java is polling for one event while there could be 100's of events in the BlockingQueue and it is very much possible that the program execution stops before the while loop can run as many times as there are log events in the queue. To fix this, I had modified the design to use `BlockingQueue.drainTo(collection)` approach and push all logs to cloudwatch at one go. Please see https://stackoverflow.com/a/48230308/829542 for implementation info. Hope this helps in enhancing the appender on your end for boxfuse environments. Thanks.

**Comment by dev-usa** [comment]
The reason I did not do a PR is because, I did not test this in the boxfuse environment. I tested with a generic version of appender (that used boxfuse's as the starting point) that has deviated so much from boxfuse's implementation. I have also exposed the configuration for the MAX_FLUSH_DELAY from CloudwatchLogsLog4J2Appender.java to the appender's config and tuned that in combination with the maxEventQueueSize for my specific needs.

About your batch size limits question, I did not change the existing behavior. The existing behavior seemed to be - if the individual log size is greater than the cloudwatch limit, the log event is ignored otherwise, it is appended to a "eventBatch" and queued for a future push. the future push happens when the desired batch size / the batch count or the "time to flush" are satisfied. This seemed reasonable for my implementation too. Thanks for giving me a starting point.
